wikitools/errors.py

```diff
@@ -0,0 +1,113 @@
+import collections
+
+from wikitools import console, link_parser
+
+
+class LinkError:
+    _colorise_fragment_only = False
+
+    def pretty(self):
+        return f'{console.blue("Note:")} {repr(self)}'
+
+    def pretty_location(self, article_path, lineno):
+        return "{}: {}".format(
+            console.yellow(":".join((article_path, str(lineno), str(self.pos)))),
+            self.link.colorise_location(fragment_only=self._colorise_fragment_only)
+        )
+
+    @property
+    def pretty_link(self):
+        return self.link.colorise_link(fragment_only=self._colorise_fragment_only)
+
+    @property
+    def pos(self):
+        return self.link.start + 1
+
+
+class MalformedLinkError(
+    LinkError,
+    collections.namedtuple('MalformedLinkError', 'link')
+):
+    """
+    An error indicating an erroneous link (for example, with several leading slashes).
+    """
+
+    link: link_parser.Link
+
+    def __repr__(self):
+        return f'Incorrect link structure (typo?): "{self.link.raw_location}"'
+
+
+class LinkNotFoundError(
+    LinkError,
+    collections.namedtuple('LinkNotFound', 'link resolved_location')
+):
+    """
+    An error indicating a missing link: a text or binary file does not exist, and there is no redirect for it.
+    """
+
+    link: link_parser.Link
+    resolved_location: str
+
+    def __repr__(self):
+        return f'"/{self.resolved_location}" was not found'
```

Comment:
```
wiki/Tournaments/3WC/2021/en.md:359:1: /wiki/shared/flag/XX.gif
Note: "/shared/flag/XX.gif" was not found

[flag_XX]: /wiki/shared/flag/XX.gif** | 5 | ![][flag_CA] Canada | [#1](https://osu.ppy.sh/community/matches/80575588) |
```

Reply:
fixed in dbb2fbf660bfcf23a6e3f4d733454a4d6290e96b

Reply:
it uh didn't get fixed